### PR TITLE
Node.js AWS Task #3 Solution

### PR DESCRIPTION
**Pull request by OHalahan**
## product-service API 
**all products**
https://4wu2eb644j.execute-api.eu-west-1.amazonaws.com/dev/products
**product by ID**
https://4wu2eb644j.execute-api.eu-west-1.amazonaws.com/dev/products/2

## Cloudfront (frontend with API from above)
**link**
https://d3ehhvbc9pzoid.cloudfront.net/
image

**Pull request into FE repository with API-related changes**
https://github.com/OHalahan/nodejs-aws-fe/pull/2

## Completed items
* 1 - product-service serverless.yml contains configuration for 2 lambda functions, API is not working at all, but YAML configuration is correct ✅
* 2 - The getProductsList OR getProductsById lambda function returns a correct response (POINT1) ✅
* 3 - The getProductsById AND getProductsList lambda functions return a correct response code (POINT2) ✅
* 4 - Your own Frontend application is integrated with product service (/products API) and products from product-service are represented on Frontend. AND POINT1 and POINT2 are done. ✅

**Additional (optional) tasks (but nice to have):**

* +1 - Async/await is used in lambda functions ✅
* +1 - ES6 modules are used for product-service implementation ✅
* +1 - Webpack is configured for product-service ✅
* +1 - SWAGGER documentation is created for product-service  ✅
** swagger.yaml is provided in the repository
** documentation can be reviewed here
** https://app.swaggerhub.com/apis/OHalahan/product-service/1.0.0 
** https://app.swaggerhub.com/apis-docs/OHalahan/product-service/1.0.0 
* +1 - Lambda handlers are covered by basic UNIT tests (NO infrastructure logic is needed to be covered) (You may use JEST) ✅
tests

* +1 - Lambda handlers (getProductsList, getProductsById) code is written not in 1 single module (file) and separated in codebase. ✅
* +1 - Main error scenarios are handled by API ("Product not found" error, try-catch blocks are used in lambda handlers). ✅

### CODE REVIEW

**Comment by develjs** [comment]
I suppose, you have not finished the integration with frontend part, because page https://d3ehhvbc9pzoid.cloudfront.net/admin/products does not show any products ;-)

File: product-service/handlers/getProductById.ts [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+
+import { HEADERS } from "../shared/constants/headers";
+
+export const getProductById = async (event: APIGatewayProxyEvent) => {
+
+    try {
+
+        const { id } = event.pathParameters || {};
+
+        if (id === null || id === undefined || id === '') {
```
Comment (develjs) [inline product-service/handlers/getProductById.ts:13]: Are you sure, it can happen? Does id = empty mean a request to `products/` url, does not it?

**Comment by OHalahan** [comment]
> I suppose, you have not finished the integration with frontend part, because page https://d3ehhvbc9pzoid.cloudfront.net/admin/products does not show any products ;-)

Haven't even checked that page.

Thank you for noticing. will fix that shortly

**Review by develjs — APPROVED** [review]

File: product-service/handlers/getProductById.ts (hunk 1 above)
Comment (OHalahan) [reply product-service/handlers/getProductById.ts:13]: That's not supposed to happen.

But taking into consideration that routing between **products/** and **products/{id}** is managed by third party (API Gateway) I've decided to add this fallback case.

**Comment by OHalahan** [comment]
> I suppose, you have not finished the integration with frontend part, because page https://d3ehhvbc9pzoid.cloudfront.net/admin/products does not show any products ;-)

fixed and deployed

**Review by Ivanko5417 — APPROVED** [review]

**Comment by MaxGaranin** [comment]
Nice work!
I think you need fix tests after adding 'count' property in Product.

**Comment by OHalahan** [comment]
> Nice work!
> I think you need fix tests after adding 'count' property in Product.

Hi, 

Definitely, will adjust tests accordingly.

Thanks :)

**Pull request merged**